packages/block-library/src/post-content/index.php

```diff
@@ -37,3 +37,21 @@ function register_block_core_post_content() {
 	);
 }
 add_action( 'init', 'register_block_core_post_content' );
+
+/**
+ * Only show the `core/post-content` block when editing templates and template parts.
+ * This is to prevent infinite recursions (post content containing a Post Content block
+ * that attempts to embed the post content it is part of).
+ *
+ * @param bool|array $allowed_block_types Array of block type slugs, or
+ *                                        boolean to enable/disable all.
+ * @param WP_Post    $post                The post resource data.
+ */
+function disallow_core_post_content_block_outside_templates( $allowed_block_types, $post ) {
+	if ( in_array( $post->post_type, array( 'wp_template', 'wp_template_part' ), true ) ) {
+		return $allowed_block_types;
+	}
+	return array_diff( $allowed_block_types, array( 'core/post-content' ) );
+}
+
+add_filter( 'allowed_block_types', 'disallow_core_post_content_block_outside_templates', 10, 2 );
```

Comment:
We might need to move this line into an action that's called upon `init`.

Reply:
Note that the filter doesn't seem to be called for the Site Editor -- I think it's only applied through Core's [`edit-form-blocks.php`](https://github.com/WordPress/WordPress/blob/6a88c91d3a2b33ed45eca9e2da33429e0f76b2a9/wp-admin/edit-form-blocks.php#L150). If we want to run it for the Site Editor, we'll need to add it to `edit-site-page.php`.